plugins/rest/src/authenticator.cpp

```diff
@@ -0,0 +1,185 @@
+  auto fb = flatbuffer<fbs::ServerState>::make(
+    std::move(chunk), flatbuffer<fbs::ServerState>::verify::yes);
+  if (!fb)
+    return fb.error();
+  if ((*fb)->server_state_type() != fbs::server_state::ServerState::v0)
+    return caf::make_error(ec::format_error, "unknown state version");
+  auto const* state = (*fb)->server_state_as_v0();
+  if (state == nullptr)
+    return caf::make_error(ec::format_error, "missing state");
+  VAST_ASSERT_CHEAP(state->auth_tokens() != nullptr);
+  using clock = std::chrono::system_clock;
+  for (auto const* token : *state->auth_tokens()) {
+    VAST_ASSERT_CHEAP(token);
+    auto token_desc = token_description{
+      .name = token->name()->str(),
+      .issued_at = clock::from_time_t(token->issued_at()),
+      .expires_at = clock::from_time_t(token->expires_at()),
+      .token = token->token()->str(),
+    };
+    tokens_.emplace_back(std::move(token_desc));
+  }
+  return {};
+}
+
+caf::expected<chunk_ptr> authenticator_state::save() {
+  flatbuffers::FlatBufferBuilder builder;
+  std::vector<
+    flatbuffers::Offset<fbs::server_state::AuthenticationTokenDescription>>
+    offsets;
+  for (auto& description : tokens_) {
+    auto name_offset = builder.CreateString(description.name);
+    auto token_offset = builder.CreateString(description.token);
+    fbs::server_state::AuthenticationTokenDescriptionBuilder token_builder(
+      builder);
+    token_builder.add_name(name_offset);
+    token_builder.add_token(token_offset);
+    token_builder.add_issued_at(
+      std::chrono::system_clock::to_time_t(description.issued_at));
+    token_builder.add_expires_at(
+      std::chrono::system_clock::to_time_t(description.expires_at));
+    offsets.push_back(token_builder.Finish());
+  }
+  auto vector_offset = builder.CreateVector(offsets);
+  fbs::server_state::v0Builder v0_builder(builder);
+  v0_builder.add_auth_tokens(vector_offset);
+  auto v0_offset = v0_builder.Finish();
+  fbs::ServerStateBuilder state_builder(builder);
+  state_builder.add_server_state_type(fbs::server_state::ServerState::v0);
+  state_builder.add_server_state(v0_offset.Union());
+  auto state_offset = state_builder.Finish();
+  fbs::FinishServerStateBuffer(builder, state_offset);
+  return chunk::make(builder.Release());
+}
+
+auto authenticator_state::generate() -> token_t {
+  auto random_bytes = std::array<char, 16>{};
+  // Use `getrandom()` for cryptographically secure random bytes.
+  while (auto error
+         = ::getrandom(random_bytes.data(), random_bytes.size(), 0)) {
+    // We should never get EFAULT or EINVAL unless our arguments
+    // are incorrect, and we already checked that the kernel
+    // supports `getrandom()`. (TODO on the last part)
+    VAST_ASSERT_CHEAP(error == EAGAIN || error == EINTR);
+  }
+  auto token = detail::base64::encode(
+    std::string_view{random_bytes.data(), random_bytes.size()});
+  // TODO: Allow the caller to pass the desired name and expiry date.
```

Comment:
Open TODO.

Reply:
I think this one is fine leave for now until we also add some ways to list and display the generated tokens.